### ISSUE TITLE
Is there really no way to limit fps?

### ISSUE THREAD

**Issue by aaratha** [not-fl3/macroquad#749]
I am compiling my game to WASM and hosting it on itch.io, and I've noticed that the speed at which it runs is completely different depending on the refresh rate of the device I'm using. It runs fine on my computer, but runs way too fast on my Android with 90hz. Is there any sort of automatic way to lock the physics process to 60hz?

**Comment by AdamSlayer** [comment]
I don't think there is any automatic way, but it should not be too hard to use `get_time()` and wait until the next physics tick for the appropriate time using `sleep()`

**Comment by aaratha** [comment]
Yeah, this seems to be the only way... I don't think sleep() works in WASM, though.

**Comment by AdamSlayer** [comment]
A better way to do this (if fps is too high) is to keep rendering without updating physics, so for example you render 2 frames for 1 physics update. And if fps is too low, you do the opposite - update physics multiple times every frame.

**Comment by aaratha** [comment]
Oh, so I could have a couple fps ranges (58-62, 68-72, 88-92), for instance, and modify the processes based on checks against those? That sounds like it could work.

**Comment by Ztry8** [comment]
**You can't use [std::thread::sleep!](https://doc.rust-lang.org/std/thread/fn.sleep.html)** 

> Puts the current thread to sleep for at least the specified amount of time.
> 
> **The thread may sleep longer than the duration specified due to scheduling specifics or platform-dependent functionality. It will never sleep less.**
> 
> This function is blocking, and should not be used in async functions.

**You can only set the swap interval.** 
> **[not-fl3](https://github.com/not-fl3) commented [on Jul 28](https://github.com/not-fl3/macroquad/issues/755#issuecomment-2254255798)**
> Just for someone who will come here from google :)
> 
> Macroquad do not limit fps, it is not doing anything with fps at all. Some platforms (browsers in particular) gives us no control over fps, some - give siomecontrol over vsync, but it can be overriten by the user in the driver settings (windows), with env variable(linux) and, I guess, performance/battery saving settings (android/ios).
> 
> swap_interval : Some(0) in Conf might be a solution, but it does not guarantee a disbled vsync, it really is super platform specific and there are varios way user can enable/disable vsync.
> https://docs.rs/miniquad/latest/miniquad/conf/struct.Platform.html#structfield.swap_interval

The swap interval is platform specific and requires additional OpenGL extensions to support it. 
According to the [ OpenGL documentation](https://www.khronos.org/opengl/wiki/Swap_Interval), swap_interval values mean: 
-1: adaptive vsync, 
0: no vsync, 
1: vsync.

**Many questions about this have been asked before, and my answer is the same for all of them.**
Here is a list of the same questions:
#755 
#380 
#360 
#170
#52

**Comment by GustavoRamos112** [comment]
````rust
use macroquad::prelude::*;

fn window_conf() -> Conf {
    Conf {
        window_title: "Uncap fps".to_owned(),
        sample_count: 0,
        platform: miniquad::conf::Platform {
          swap_interval: Some(0), // <- this line: 0: unlimited fps, 1: 60 fps, 2: 30 fps, 3: 15 fps
          ..Default::default()
        },
        window_resizable: true,
        window_width: 800,
        window_height: 600,
        ..Default::default()
    }
}

#[macroquad::main(window_conf)]
async fn main() {
  //? main loop
  loop {
    clear_background(BLACK);

    draw_text(
      format!("{:.4} s/{:.0} fps", get_frame_time(), 1.0/(get_frame_time() as f32)).as_str(), 
      0.0, 13.0, 20.0, WHITE,
    );

    next_frame().await
  }
}
````

**Comment by afrizalhasbi** [comment]
@GustavoRamos112 the `swap_interval` method is not reliable as per macroquad's own documentation. Using it i get a pretty wide fps distribution.

I managed to get an exact FPS with this method:

```rust
    let fps_target = 60.0;
    let frame_dur = Duration::from_secs_f64(1.0 / fps_target);
    let mut next_tick = Instant::now();

    loop {
        // game logic ...

        next_tick += frame_dur;
        let now = Instant::now();
        if next_tick > now {
            spin_sleep::sleep(next_tick - now); // spin_sleep crate; might be more reliable than std::thread::sleep
        } else {
            next_tick = now; // catch up after a slow frame
        }

        next_frame().await;
    }
```

The fps is reliably around 59-60 with this method, at least on my linux machine.

So far, I have not experienced any blocking problems by the sleep.

May not work in WASM, haven't tried.
